metrics/resource_view_test.go

```diff
@@ -0,0 +1,639 @@
+	"knative.dev/pkg/metrics/metricskey"
+	//_ "knative.dev/pkg/metrics/testing"
+)
+
+var (
+	r               = resource.Resource{Labels: map[string]string{"foo": "bar"}}
+	NamespaceTagKey = tag.MustNewKey(metricskey.LabelNamespaceName)
+	ServiceTagKey   = tag.MustNewKey(metricskey.LabelServiceName)
+	ConfigTagKey    = tag.MustNewKey(metricskey.LabelConfigurationName)
+	RevisionTagKey  = tag.MustNewKey(metricskey.LabelRevisionName)
+)
+
+func TestRegisterResourceView(t *testing.T) {
+	meter := meterExporterForResource(&r).m
+
+	m := stats.Int64("testView_sum", "", stats.UnitDimensionless)
+	view := view.View{Name: "testView", Measure: m, Aggregation: view.Sum()}
+	err := RegisterResourceView(&view)
+
+	if err != nil {
+		t.Errorf("RegisterResourceView with error %v.", err)
+	}
+
+	viewToFind := defaultMeter.m.Find("testView")
+	if viewToFind == nil || viewToFind.Name != "testView" {
+		t.Errorf("Registered view should be found in default meter, instead got %v", viewToFind)
+	}
+
+	viewToFind = meter.Find("testView")
+	if viewToFind == nil || viewToFind.Name != "testView" {
+		t.Errorf("Registered view should be found in new meter, instead got %v", viewToFind)
+	}
+}
+
+func TestOptionForResource(t *testing.T) {
+	option, err1 := optionForResource(&r)
+	if err1 != nil {
+		t.Errorf("Should succeed getting option, instead got error %v", err1)
+	}
+	optionAgain, err2 := optionForResource(&r)
+	if err2 != nil {
+		t.Errorf("Should succeed getting option, instead got error %v", err2)
+	}
+
+	if fmt.Sprintf("%v", optionAgain) != fmt.Sprintf("%v", option) {
+		t.Errorf("Option for the same resource should not be recreated, instead got %v and %v", optionAgain, option)
+	}
+}
+
+type testExporter struct {
+	id string
+}
+
+func (fe *testExporter) ExportView(vd *view.Data) {}
+func (fe *testExporter) Flush()                   {}
+func TestSetFactor(t *testing.T) {
+	fakeFactory := func(rr *resource.Resource) (view.Exporter, error) {
+		if rr == nil {
+			return &testExporter{}, nil
+		}
+
+		return &testExporter{id: rr.Labels["id"]}, nil
+	}
+
+	resource123 := r
+	resource123.Labels["id"] = "123"
+
+	setFactory(fakeFactory)
+	// Create the new meter and apply the factory
+	_, err := optionForResource(&resource123)
+	if err != nil {
+		t.Errorf("Should succeed getting option, instead got error %v", err)
+	}
+
+	// Now get the exporter and verify the id
+	me := meterExporterForResource(&resource123)
+	e := me.e.(*testExporter)
+	if e.id != "123" {
+		t.Errorf("Expect id to be 123, instead got %v", e.id)
+	}
+
+	resource456 := r
+	resource456.Labels["id"] = "456"
+	// Create the new meter and apply the factory
+	_, err = optionForResource(&resource456)
+	if err != nil {
+		t.Errorf("Should succeed getting option, instead got error %v", err)
+	}
+
+	me = meterExporterForResource(&resource456)
+	e = me.e.(*testExporter)
+	if e.id != "456" {
+		t.Errorf("Expect id to be 456, instead got %v", e.id)
+	}
+}
+
+func TestResourceAsString(t *testing.T) {
```

Comment:
`resourceFromString` is fairly straightforward, and is needed/used by a bunch of the tests indirectly. `resourceAsString` is more subtle (and, in fact, was the source of some of the test flakiness in `TestMetricsExport`) because _the order of go map iteration is undefined_ and may vary between and within program executions (so two Resources with the same keys and values in the map may iterate over the keys in different orders).